ProjectSidewalk/SidewalkWebpage
# Resolves #1240 Adds clustering of labels into attributes

### Testing Instructions
NOTE: You will probably end up needing to install a bunch of Python packages :)

1. Just try to run the code and make sure that it all compiles and there are no problems with the database evolutions.
1. Check that the `accessibility_feature` and `sidewalk_edge_accessibility_feature` tables are no longer in the database and that the `user_clustering_session`, `user_attribute`, `user_attribute_label`, `global_clustering_session`, `global_attribute`, and `global_attribute_user_attribute` tables have been added to the database.
1. Check that the `label_type` table now has an 8th entry for the `Problem` label type.
1. Make sure you are not logged in.
1. Go to the endpoint [http://localhost:9000/clustering](http://localhost:9000/clustering); you should be redirected to the landing page because you are not signed in as an admin.
1. Log in as an admin.
1. Go to the endpoint [http://localhost:9000/clustering](http://localhost:9000/clustering), and a simple clustering page should appear, that looks like this:
![clustering_page](https://user-images.githubusercontent.com/6518824/41859665-43fcbef0-7862-11e8-8de1-790f995e46a9.png)
1. Click on the "Single AND Multi User Clustering" button. You should see an error in the window where you ran `sbt run`, complaining about not being able to read a keyfile, or a key not matching.
1. I will send you a file called `special_api_key.txt`, which you should put in the project's root directory.
1. Now click on the button again. This will start the clustering process, which could take up to around 30 minutes, depending on the size of the database dump you are working with and how fast your computer is.
1. The text "Running..." should appear below the buttons on that page, like so:
    ![running_text](https://user-images.githubusercontent.com/6518824/41859765-8181f074-7862-11e8-91f0-53f4637d5b90.png)
1. In the terminal where you ran the `sbt run` command, progress is being printed out (I will comment out or remove this before merging, but it is there to aid in testing). First, the single user clustering is run (anon users, then registered users; it will switch about halfway through), followed by multi-user clustering by region. Images below (in the first image, I covered the anon user IP addresses, but you will see those in the terminal).
    ![user_progress](https://user-images.githubusercontent.com/6518824/41860061-47fb4976-7863-11e8-88e7-2c2eccda9ead.png)
    ![region_progress](https://user-images.githubusercontent.com/6518824/41864790-b61a7ba4-7870-11e8-89ab-5d7a29ae20ee.png)
1. When the code finishes running, you should see the following on the clustering page (although I believe that you are using a smaller data dump, so these numbers should be smaller for you). Make sure that the number of labels is larger than the number of user attributes, which is larger than the number of global attributes.
    ![clustering_results](https://user-images.githubusercontent.com/6518824/41865211-082817ca-7872-11e8-87a8-1ee1166ed726.png)
1. Check that these numbers match the number of records that are now in the `user_attribute_label`, `user_attribute`, and `global_attribute` tables for number of labels, user attributes, and global attributes, respectively.
1. Write down the number of records in each of those tables.
1. Now try clicking on just the "Multi User Clustering" button. You should see the multi-user clustering re running in the terminal now.
1. When that finishes, check that the number of records in each table is the same as before you re ran multi-user clustering.
1. Now try clicking the "Single User Clustering" button. The number of records in the tables should again be the same, except that the `global_attribute` and `global_attribute_user_attribute` tables should be empty. This is because we deleted the records in the `user_attribute` table and re-ran it. The `global_attribute` entries were based off of these `user_attribute` entries, so we had to delete the global attributes as well.
1. Check out the `develop` branch and recompile the code. You just need to make sure that the code compiles here. This check is because I made some changes to the database in the files `conf/evolutions/default/*.sql`. In these files, one has to specify the changes you are making (the "ups"), as well as how to undo the changes (the "downs"). When you compile the code with those evolutions, the ups are automatically applied, and when you compile other code afterwards, the downs are automatically applied. Often, the downs are not specified correctly, which causes errors when you move from that branch back to develop. This is why I want you to verify that the develop code compiles!

Summary of changes
1. Added new tables to the database (`user_clustering_session`, `user_attribute`, `user_attribute_label`, `global_clustering_session`, `global_attribute`, and `global_attribute_user_attribute`)
2. Added single-user clustering, which fills the `user_attribute` table with the clusters, while the `user_attribute_label` table links those clusters to the labels that make up those clusters.
3. Added multi-user clustering, which fills the `global_attribute` table with the clusters, while the `global_attribute_user_attribute` table links those clusters to the user attributes that make up those clusters.
4. Added a new endpoint `/clustering` that requires one to be logged in as an admin to access. It has buttons that can start the clustering process (or you can just do single- or multi-user clustering by itself).
5. Removed two old tables from the database that were never used (`accessibility_feature` and `sidewalk_edge_accessibility_feature`).

More detailed description of how the code actually works:
1. From the /clustering page, someone clicks the "Single AND Multi User Clustering" button.
2. This calls the `/runClustering?clusteringType=both` endpoint.
3. The data in all the tables listed at the top of this comment are deleted (partial updating of the data will be done in the future).
4. We compute which users are "good" users -- having labeling frequency above the 3.75 labels per 100 meters threshold.
5. For a single user, we run the `label_clustering.py` Python script, sending the user id as a parameter. We also read an authentication key from a file and send that as a parameter.
6. This script calls the `/userLabelsToCluster` endpoint (sending the authentication key as a parameter). This endpoint sends back all the labels placed by that user.
7. The script performs clustering on those labels. Each label type is clustered separately.
8. The script finally posts the results of the clustering to the `/singleUserClusteringResults` endpoint (sending authentication key as param again).
9. This endpoint adds the clustering data to the `user_clustering_session`, `user_attribute`, and `user_attribute_label` tables.
10. Steps 5 through 9 are repeated for each user individually.
11. For a single region (i.e., neighborhood), we run the `label_clustering.py` Python script again, sending the region id as a parameter, and key as a param (same as in step 5).
12. This script calls the `/clusteredLabelsInRegion` endpoint (sending the authentication key as a parameter). This endpoint sends back all the user attribute (i.e., user clusters) by any (high quality) users, where the user attribute's lat-lng is in that region.
13. The script performs clustering on those attributes. Each label type is clustered separately again. The max clustering distance is increased for the multi-user clustering, but we make sure that a single user's attributes are never clustered together.
14. The script finally posts the results of the clustering to the `/multiUserClusteringResults` endpoint (sending authentication key as param again).
15. This endpoint adds the clustering data to the `global_clustering_session`, `global_attribute`, and `global_attribute_user_attribute` tables.
